### PR TITLE
Add managedosversion finalizer

### PR DESCRIPTION
**Pull request by anmazzotti**
This PR is a pre-requisite to #759 

A new controller has been added to manage the ManagedOSVersion finalizer. 
This will prevent deletion of ManagedOSVersion that are still referenced by any ManagedOSImage.

To my knowledge there is no other direct reference to a ManagedOSVersion. The UI uses a cached list of versions when creating a new SeedImage, but this is presentation layer only, there's no reference to a ManagedOSVersion from a SeedImage.

### CODE REVIEW

File: api/v1beta1/machineinventory_types.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-var (
+const (
```
Comment (anmazzotti) [inline api/v1beta1/machineinventory_types.go:24]: Very random change. Not related to this PR.

File: api/v1beta1/managedosversion_types.go [hunk 2]
```diff
@@ -59,9 +59,7 @@ type ManagedOSVersionSpec struct {
 	UpgradeContainer *upgradev1.ContainerSpec `json:"upgradeContainer,omitempty"`
 }
 
-type ManagedOSVersionStatus struct {
-	fleetv1.BundleStatus `json:""` //nolint
-}
+type ManagedOSVersionStatus struct{}
```
Comment (anmazzotti) [inline api/v1beta1/managedosversion_types.go:62]: I tested on main branch by creating a standalone ManagedOSVersion, and also by using a ManagedOSVersionChannel to create some more.
None of them had any `status`, so this is not in use.

**Comment by anmazzotti** [comment]
Also accidentally found how it looks like when the finalizer is preventing the deletion.
So even without UI changes, this is looking good.

scheduled_for_deletion

File: controllers/managedosversion_controller.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+func (r *ManagedOSVersionReconciler) ManagedOSImageToManagedOSVersion(ctx context.Context, obj client.Object) []ctrl.Request {
+	logger := ctrl.LoggerFrom(ctx)
+	logger.Info("Enqueueing ManagedOSVersion reconciliation from ManagedOSImage")
+
+	requests := []ctrl.Request{}
+
+	// Verify we are actually handling a ManagedOSImage object
+	managedOSImage, ok := obj.(*elementalv1.ManagedOSImage)
+	if !ok {
+		logger.Error(errors.New("Enqueueing error"), fmt.Sprintf("Expected a ElementalHost object, but got %T", obj))
```
Comment (fgiudici) [inline controllers/managedosversion_controller.go:65]:
```suggestion
		logger.Error(errors.New("Enqueueing error"), fmt.Sprintf("Expected a ManagedOSImage object, but got %T", obj))
```

File: controllers/managedosversion_controller.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		return []ctrl.Request{}
+	}
+
+	// Check the ManagedOSImage was referencing any ManagedOSVersion
+	if len(managedOSImage.Spec.ManagedOSVersionName) > 0 {
```
Comment (fgiudici) [inline controllers/managedosversion_controller.go:70]: 🤔 is this to clean-up stale ManagedOSVersion resources when the ManagedOSImage is deleted?

**Review by fgiudici — APPROVED** [review]
Nice PR! just found a wrong resource mentioned in an error message

File: controllers/managedosversion_controller.go (hunk 3 above)
Comment (anmazzotti) [reply controllers/managedosversion_controller.go:65]: Too much copy pasting :D

File: controllers/managedosversion_controller.go (hunk 4 above)
Comment (anmazzotti) [reply controllers/managedosversion_controller.go:70]: Yes. 

The ManagedOSVersionController will **not** requeue after trying to remove the finalizer the first time, so we need to watch ManagedOSImages to check for changes. Makes it pretty efficient.

File: api/v1beta1/managedosversion_types.go (hunk 2 above)
Comment (anmazzotti) [reply api/v1beta1/managedosversion_types.go:62]: @davidcassany already suggested to purge this status :D
I can confirm is not in use, so it's safe.

**Pull request merged**